opm/simulators/flow/partitionCells.cpp

```diff
@@ -357,6 +363,122 @@ partitionCellsZoltan(const int                                      num_domains,
+    {
+        auto groupedCells = std::vector<int>{};
+        auto startPtr = std::vector<int>(comm.size() + 1, 0);
+
+        if (comm.rank() == 0) {
+            // VertexID = std::vector<int>::size_type
+            // TrackCompressedIdx = false
+            // PermitSelfConnections = true (e.g., process 0 has row 0).
+            auto cellGroup = Opm::utility::CSRGraphFromCoordinates<
+                std::vector<int>::size_type, false, true
+            >{};
+
+            const auto numCells = rawInput.size() / 3;
+            for (auto cell = 0*numCells; cell < numCells; ++cell) {
+                cellGroup.addConnection(rawInput[3*cell + 0], cell);
+            }
+            cellGroup.compress(comm.size());
+
+            const auto& inputIx = cellGroup.columnIndices();
+            groupedCells.reserve(2 * inputIx.size());
+
+            for (const auto& cell : inputIx) {
+                const auto* cellVal = &rawInput[3*cell + 0];
+                groupedCells.push_back(cellVal[1]); // Cartesian index
+                groupedCells.push_back(cellVal[2]); // Block ID
+            }
+
+            const auto& start = cellGroup.startPointers();
+            std::copy(start.begin(), start.end(), startPtr.begin());
+        }
+
+        comm.broadcast(startPtr.data(), comm.size() + 1, 0);
+
+        // We're scattering two ints per cell
+        for (auto& startIx : startPtr) {
+            startIx *= 2;
+        }
+
+        auto sendLength = std::vector<int>(comm.size());
+        std::adjacent_difference(startPtr.begin() + 1,
+                                 startPtr.end(),
+                                 sendLength.begin());
+
+        auto perProcessInput = std::vector<int>(sendLength[comm.rank()], 0);
+        comm.scatterv(groupedCells.data(),
+                      sendLength.data(),
+                      startPtr.data(),
+                      perProcessInput.data(),
+                      static_cast<int>(perProcessInput.size()),
+                      0);
+
+        return perProcessInput;
+    }
+
+    template <class GridView, class Element>
+    std::vector<int>
+    scatterPartition(const std::vector<int>&                        rawInput,
+                     const GridView&                                grid_view,
+                     const Opm::ZoltanPartitioningControl<Element>& zoltan_ctrl)
+    {
+        auto g2l = std::unordered_map<int,int>{};
+        {
+            auto localCell = 0;
+
+            for (const auto& cell : elements(grid_view, Dune::Partitions::interior)) {
+                const auto globIx = zoltan_ctrl.local_to_global(zoltan_ctrl.index(cell));
+                g2l.insert_or_assign(globIx, localCell++);
+            }
+        }
+
+        const auto myPartitionInput = groupInputByProcessID(rawInput, grid_view.comm());
+
+        auto partition = std::vector<int>(g2l.size(), -1);
+        for (auto i = 0*myPartitionInput.size(); i < myPartitionInput.size(); i += 2) {
```

Comment:
> More readable to increment by 1 and do 2*i inside the loop.

Maybe, but then we can't use `myPartitionInput.size()` as the upper bound.  It would be something like `myPartitionInput.size() / 2` and at that point it's not clear to me which is better.